### PR TITLE
feat: support `--format` for `oras pull`

### PR DESCRIPTION
**Pull request by qweeah**
**What this PR does / why we need it**:
This PR
1) Adds support for `--format [json|<go_template>]` for `pull` command according to #1199
2) Refactor status/metadata output

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Related to #638

**Please check the following list**:
- [x]  Does the affected code have corresponding tests, e.g. unit test, E2E test?
- [ ]  Does this change require a documentation update?
- [ ]  Does this introduce breaking changes that would require an announcement or bumping the major version?
- [ ]  Do all new files have an appropriate license header?

### CODE REVIEW

File: cmd/oras/root/pull.go [hunk 1]
```diff
@@ -134,65 +136,45 @@ func runPull(cmd *cobra.Command, opts *pullOptions) error {
...
+	desc, layerSkipped, files, err := doPull(ctx, src, dst, copyOptions, statusHandler, opts)
 	if err != nil {
 		if errors.Is(err, file.ErrPathTraversalDisallowed) {
 			err = fmt.Errorf("%s: %w", "use flag --allow-path-traversal to allow insecurely pulling files outside of working directory", err)
 		}
 		return err
 	}
 
-	// suggest oras copy for pulling layers without annotation
-	outWriter := cmd.OutOrStdout()
```
Comment (shizhMSFT) [inline cmd/oras/root/pull.go:146]: `cmd.OutOrStdout()` is not migrated to the new model

File: cmd/oras/internal/display/status/interface.go [hunk 2]
```diff
@@ -30,3 +33,22 @@ type PushHandler interface {
...
 // AttachHandler handles text status output for attach command.
 type AttachHandler PushHandler
+
+// PullHandler handles status output for pull command.
+type PullHandler interface {
+	// TrackTarget returns a tracked target.
+	// If no TTY is available, it returns the original target.
+	TrackTarget(gt oras.GraphTarget) (oras.GraphTarget, error)
+	// StopTracking stops tracking the target if available.
+	StopTracking()
```
Comment (shizhMSFT) [inline cmd/oras/internal/display/status/interface.go:43]: Should it just be `io.Closer`?

File: cmd/oras/internal/display/status/interface.go [hunk 3]
```diff
@@ -30,3 +33,22 @@ type PushHandler interface {
...
+	// OnNodeProcessing is called when processing a manifest.
+	OnNodeProcessing(desc ocispec.Descriptor) error
+	// OnNodeDownloading is called before downloading a node.
+	OnNodeDownloading(desc ocispec.Descriptor) error
+	// OnNodeDownloaded is called after a node is downloaded.
+	OnNodeDownloaded(desc ocispec.Descriptor) error
+	// OnNodeRestored is called after a deduplicated node is restored.
+	OnNodeRestored(printed *sync.Map, desc ocispec.Descriptor) error
+	// OnNodeSkipped is called when a node is skipped.
+	OnNodeSkipped(printed *sync.Map, desc ocispec.Descriptor) error
```
Comment (shizhMSFT) [inline cmd/oras/internal/display/status/interface.go:53]: `printed *sync.Map` related logic should not be there.

File: cmd/oras/internal/display/metadata/interface.go [hunk 4]
```diff
@@ -30,3 +31,9 @@ type PushHandler interface {
 type AttachHandler interface {
 	OnCompleted(opts *option.Target, root, subject ocispec.Descriptor) error
 }
+
+// PullHandler handles metadata output for attach events.
+type PullHandler interface {
+	// OnCompleted is called when the pull cmd execution is completed.
+	OnCompleted(opts *option.Target, desc ocispec.Descriptor, layerSkipped bool, _ []model.File) error
+}
```
Comment (shizhMSFT) [inline cmd/oras/internal/display/metadata/interface.go:39]: Should the `OnCompleted` method be split to `OnCompleted` and `OnFilePulled`?

**Review by shizhMSFT — CHANGES_REQUESTED** [review]

File: cmd/oras/internal/display/metadata/interface.go (hunk 4 above)
Comment (qweeah) [reply cmd/oras/internal/display/metadata/interface.go]: Why we need to split the output?

Comment (shizhMSFT) [reply cmd/oras/internal/display/metadata/interface.go]: It is to stream the metadata of the file pulling.

Comment (qweeah) [reply cmd/oras/internal/display/metadata/interface.go]: Refactored.

File: cmd/oras/internal/display/metadata/interface.go [hunk 5]
```diff
@@ -30,3 +30,13 @@ type PushHandler interface {
 type AttachHandler interface {
 	OnCompleted(opts *option.Target, root, subject ocispec.Descriptor) error
 }
+
+// PullHandler handles metadata output for pull events.
+type PullHandler interface {
+	// OnLayerSkipped is called when a layer is skipped.
+	OnLayerSkipped()
```
Comment (shizhMSFT) [inline cmd/oras/internal/display/metadata/interface.go:37]: nit: should it take a `Descriptor` although the descriptor is not actually used?

File: cmd/oras/internal/display/metadata/model/pull.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+			Ref: digestReference,
+		},
+		Files: pulled,
+	}
+}
+
+// Pulled records all pulled files.
+type Pulled struct {
+	lock  sync.Mutex
+	Files []File
```
Comment (shizhMSFT) [inline cmd/oras/internal/display/metadata/model/pull.go:67]: In terms encapsulation, `Files` not should be exposed.

File: cmd/oras/internal/display/metadata/model/pull.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+// Add adds a pulled file.
+func (f *Pulled) Add(name string, outputDir string, desc ocispec.Descriptor, descPath string) {
```
Comment (shizhMSFT) [inline cmd/oras/internal/display/metadata/model/pull.go:71]: code style: naming
```suggestion
func (p *Pulled) Add(name string, outputDir string, desc ocispec.Descriptor, descPath string) {
```

File: cmd/oras/internal/display/metadata/model/pull.go [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+// newFile creates a new file metadata.
+func newFile(name string, outputDir string, desc ocispec.Descriptor, descPath string) File {
+	path := name
+	if !filepath.IsAbs(name) {
+		// ignore error since it's successfully written to file store
+		path, _ = filepath.Abs(filepath.Join(outputDir, name))
+	}
+	if desc.Annotations[file.AnnotationUnpack] == "true" {
+		path += string(filepath.Separator)
```
Comment (shizhMSFT) [inline cmd/oras/internal/display/metadata/model/pull.go:41]: What if `name` is in Unix style but `Separator` is in Windows style?

File: cmd/oras/internal/display/status/interface.go [hunk 9]
```diff
@@ -30,3 +33,21 @@ type PushHandler interface {
...
+	OnNodeProcessing(desc ocispec.Descriptor) error
+	// OnNodeDownloading is called before downloading a node.
+	OnNodeDownloading(desc ocispec.Descriptor) error
+	// OnNodeDownloaded is called after a node is downloaded.
+	OnNodeDownloaded(desc ocispec.Descriptor) error
+	// OnNodeRestored is called after a deduplicated node is restored.
+	OnNodeRestored(desc ocispec.Descriptor) error
+	// OnNodeSkipped is called when a node is skipped.
+	OnNodeSkipped(desc ocispec.Descriptor) error
+	io.Closer
```
Comment (shizhMSFT) [inline cmd/oras/internal/display/status/interface.go:52]: nit: code style: could make `io.Closer` the first field?

File: cmd/oras/internal/display/status/utils.go [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package status
+
+import v1 "github.com/opencontainers/image-spec/specs-go/v1"
```
Comment (shizhMSFT) [inline cmd/oras/internal/display/status/utils.go:18]: nit:
```suggestion
import ocispec "github.com/opencontainers/image-spec/specs-go/v1"
```

File: cmd/oras/internal/display/metadata/interface.go (hunk 5 above)
Comment (shizhMSFT) [reply cmd/oras/internal/display/metadata/interface.go:37]: nit: should it return an error?

File: cmd/oras/root/pull.go [hunk 11]
```diff
@@ -278,54 +247,20 @@ func doPull(ctx context.Context, src oras.ReadOnlyTarget, dst oras.GraphTarget,
...
-	return desc, layerSkipped.Load(), err
-}
-
-// generateContentKey generates a unique key for each content descriptor, using
-// its digest and name if applicable.
-func generateContentKey(desc ocispec.Descriptor) string {
-	return desc.Digest.String() + desc.Annotations[ocispec.AnnotationTitle]
-}
-
-func printOnce(printed *sync.Map, s ocispec.Descriptor, msg string, verbose bool, dst any) error {
```
Comment (shizhMSFT) [inline cmd/oras/root/pull.go:309]: I think we can still keep it but change to
```go
func notifyOnce(notified *sync.Map, s ocispec.Descriptor, notify func(ocispec.Descriptor) error) error {
```

File: cmd/oras/internal/display/metadata/interface.go [hunk 12]
```diff
@@ -30,3 +30,13 @@ type PushHandler interface {
...
+	// OnFilePulled is called after a file is pulled.
+	OnFilePulled(name string, outputDir string, desc ocispec.Descriptor, descPath string)
```
Comment (shizhMSFT) [inline cmd/oras/internal/display/metadata/interface.go:39]: nit: should it return an error?

File: cmd/oras/internal/display/metadata/interface.go (hunk 5 above)
Comment (qweeah) [reply cmd/oras/internal/display/metadata/interface.go:37]: Added descriptor as parameter. Why an error is needed? Currently there is not output required on layer skipped events.

File: cmd/oras/internal/display/metadata/model/pull.go (hunk 6 above)
Comment (qweeah) [reply cmd/oras/internal/display/metadata/model/pull.go:67]: If it's not public, go template cannot select it.

File: cmd/oras/internal/display/metadata/model/pull.go (hunk 8 above)
Comment (qweeah) [reply cmd/oras/internal/display/metadata/model/pull.go:41]: According to my test result, unix `Separator` in the `name` has already be converted to Windows' automatically before appending.

File: cmd/oras/internal/display/metadata/model/pull.go (hunk 6 above)
Comment (qweeah) [reply cmd/oras/internal/display/metadata/model/pull.go]: Added getter func

File: cmd/oras/internal/display/metadata/json/pull.go [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+
+// PullHandler handles JSON metadata output for pull events.
+type PullHandler struct {
+	path   string
+	pulled model.Pulled
+	out    io.Writer
+}
+
+// OnLayerSkipped implements metadata.PullHandler.
+func (ph *PullHandler) OnLayerSkipped(ocispec.Descriptor) error { return nil }
```
Comment (shizhMSFT) [inline cmd/oras/internal/display/metadata/json/pull.go:35]: nit: code style
```suggestion
func (ph *PullHandler) OnLayerSkipped(ocispec.Descriptor) error {
    return nil
}
```

File: cmd/oras/internal/display/metadata/template/pull.go [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+	return parseAndWrite(ph.out, model.NewPull(ph.path+"@"+desc.Digest.String(), ph.pulled.Files()), ph.template)
+}
+
+// OnFilePulled implements metadata.PullHandler.
+func (ph *PullHandler) OnFilePulled(name string, outputDir string, desc ocispec.Descriptor, descPath string) error {
+	return ph.pulled.Add(name, outputDir, desc, descPath)
+}
+
+// OnLayerSkipped implements metadata.PullHandler.
+func (ph *PullHandler) OnLayerSkipped(ocispec.Descriptor) error { return nil }
```
Comment (shizhMSFT) [inline cmd/oras/internal/display/metadata/template/pull.go:46]: nit: code style
```suggestion
func (ph *PullHandler) OnLayerSkipped(ocispec.Descriptor) error {
    return nil
}
```

File: cmd/oras/internal/display/status/utils.go [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+
+import ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+
+// GenerateContentKey generates a unique key for each content descriptor, using
+// its digest and name if applicable.
+func GenerateContentKey(desc ocispec.Descriptor) string {
+	return desc.Digest.String() + desc.Annotations[ocispec.AnnotationTitle]
+}
+
+const (
```
Comment (shizhMSFT) [inline cmd/oras/internal/display/status/utils.go:26]: Request documentation

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**